File: python/2015/16/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

tape = dict({
    "children": 3,
    "cats": 7,
    "samoyeds": 2,
    "pomeranians": 3,
    "akitas": 0,
    "vizslas": 0,
    "goldfish": 5,
    "trees": 3,
    "cars": 2,
    "perfumes": 1,
})

def read_input():
    return [(a,int(b),c,int(d),e,int(f)) for a,b,c,d,e,f in u.read_match(file, ".* (.*): (.*), (.*): (.*), (.*): (.*)")]

def part1():
    A = read_input()
    k = 0
    for i,a in enumerate(A):
        if tape[a[0]] == a[1] and tape[a[2]] == a[3] and tape[a[4]] == a[5]:
            k = i+1
            break
    print("Part 1:",k)

def t(a,v):
    k = tape[a]
    if a in ["cats", "trees"]:
        return k < v
    elif a in ["pomeranians", "goldfish"]:
        return k > v
    else:
        return k == v

def part2():
    A = read_input()
    k = -1
    for i,a in enumerate(A):
        if t(a[0], a[1]) and t(a[2], a[3]) and t(a[4], a[5]):
            k = i+1
            break
    print("Part 2:",k)

part1()
part2()

File: c/2015/16/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

struct val {
    char name[15];
    int8_t val;
};

struct val AUNT[10] = {{.name = "children", .val = 3}, {.name="cats", .val=7}, {.name="samoyeds", .val=2}, {.name="pomeranians", .val=3}, {.name="akitas", .val=0}, {.name="vizslas", .val=0}, {.name="goldfish", .val=5}, {.name="trees", .val=3}, {.name="cars", .val=2}, {.name="perfumes", .val=1}};

bool t(struct val aunt) {
    for (int8_t i = 0; i < 10; i++) {
        if (!strcmp(aunt.name, AUNT[i].name)) {
            return aunt.val == AUNT[i].val;
        }
    }
    return false;
}

bool tt(struct val aunt) {
    for (int8_t i = 0; i < 10; i++) {
        if (!strcmp(aunt.name, AUNT[i].name)) {
            if (!strcmp(aunt.name, "cats") || !strcmp(aunt.name, "trees")) {
                return aunt.val > AUNT[i].val;
            } else if (!strcmp(aunt.name, "pomeranians") || !strcmp(aunt.name, "goldfish")) {
                return aunt.val < AUNT[i].val;
            } else {
                return aunt.val == AUNT[i].val;
            }
        }
    }
    return false;
}

void part1() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " :,");
    int64_t k = 0;
    struct val aunt[500][3];
    for (int16_t i = 0; i < in->length; i++) {
        for (int16_t j = 0; j < 3; j++) {
            strcpy(aunt[i][j].name, in->strings[i].strings[2 + 2*j]);
            aunt[i][j].val = atoi(in->strings[i].strings[3 + 2*j]);
        }
        if (t(aunt[i][0]) && t(aunt[i][1]) && t(aunt[i][2])) {
            k = i+1;
            break;
        }
    }
    printf("Part 1: %ld\n", k);
    freeArrMulString(in);
}

void part2() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " :,");
    int64_t k = 0;
    struct val aunt[500][3];
    for (int16_t i = 0; i < in->length; i++) {
        for (int16_t j = 0; j < 3; j++) {
            strcpy(aunt[i][j].name, in->strings[i].strings[2 + 2*j]);
            aunt[i][j].val = atoi(in->strings[i].strings[3 + 2*j]);
        }
        if (tt(aunt[i][0]) && tt(aunt[i][1]) && tt(aunt[i][2])) {
            k = i+1;
            break;
        }
    }
    printf("Part 2: %ld\n", k);
    freeArrMulString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
